source/simulator/helper_functions.cc

```diff
@@ -985,49 +985,144 @@ namespace aspect
+                      points_values[k_index](2) = z(0);
+                    }
+                }
+            }
+          const unsigned int n_q_points_121 = n_q_points_1 * n_q_points_2 * n_q_points_1;
+          //make quadrature points combination 112
+          for ( unsigned int i=0; i < n_q_points_1 ; i++)
+            {
+              const Point <1>  x = quadrature_formula_1.point(i);
+              for ( unsigned int j=0; j < n_q_points_1 ; j++)
+                {
+                  const Point <1>  y = quadrature_formula_1.point(j);
+                  for ( unsigned int k=0; k < n_q_points_2 ; k++)
+                    {
+                      const unsigned int k_index =
+                        n_q_points_121 + i * n_q_points_1 * n_q_points_2 + j * n_q_points_2 + k;
+                      const Point <1>  z = quadrature_formula_2.point(k);
+                      points_values[k_index](0) = x(0);
+                      points_values[k_index](1) = y(0);
+                      points_values[k_index](2) = z(0);
+                    }
+                }
+            }
+          //make quadrature points combination 211
+          for ( unsigned int i=0; i < n_q_points_2 ; i++)
+            {
+              const Point <1>  x = quadrature_formula_2.point(i);
+              for ( unsigned int j=0; j < n_q_points_1 ; j++)
+                {
+                  const Point <1>  y = quadrature_formula_1.point(j);
+                  for ( unsigned int k=0; k < n_q_points_1 ; k++)
+                    {
+                      const unsigned int k_index =
+                        2 * n_q_points_121 + i * n_q_points_2 * n_q_points_1 + j * n_q_points_1 + k;
+                      const Point <1>  z = quadrature_formula_1.point(k);
+                      points_values  [k_index](0) = x(0);
+                      points_values  [k_index](1) = y(0);
+                      points_values  [k_index](2) = z(0);
+                    }
+                }
+            }
+          break;
+        }
+
+        default:
+          Assert (false, ExcNotImplemented());
+      }
+    Quadrature<dim> quadrature_formula(points_values);
+
+    // Quadrature rules only used for the numerical integration for better accuracy
+    const QGauss<dim> quadrature_formula_0 (advection_field.is_temperature() ?
+                                            parameters.temperature_degree+1 :
+                                            parameters.composition_degree+1);
+
+    const unsigned int n_q_points_0 = quadrature_formula_0.size();
+
+    // fe values for points evalution
+    FEValues<dim> fe_values (mapping,
+                             finite_element,
+                             quadrature_formula,
+                             update_values   |
+                             update_quadrature_points);
+    std::vector<double> values (n_q_points);
+    // fe values for numerical integration, which points is 1/dim of the total points above
```

Comment:
I don't think I understand the second half of the comment. Do you mean "..., with a number of quadrature points that is equal to 1/dim times the number of total points above"?